src/hotspot/share/gc/shenandoah/shenandoahScanRemembered.hpp

```diff
@@ -0,0 +1,945 @@
+//           of memory pertaining to other clusters.
+//
+//      Though an initial division of labor between marking threads
+//      may assign equal numbers of clusters to be scanned by each
+//      thread, it should be expected that some threads will finish
+//      their assigned work before others.  Therefore, some amount
+//      of the full remembered set scanning effort should be held
+//      back and assigned incrementally to the threads that end up with
+//      excess capacity.  Consider the following strategy for dividing
+//      labor:
+//
+//        1. Assume there are 8 marking threads and 1024 remembered
+//           set clusters to be scanned.
+//        2. Assign each thread to scan 64 clusters.  This leaves
+//           512 (1024 - (8*64)) clusters to still be scanned.
+//        3. As the 8 server threads complete previous cluster
+//           scanning assignments, issue each of the next 8 scanning
+//           assignments as units of 32 additional cluster each.
+//           In the case that there is high variance in effort
+//           associated with previous cluster scanning assignments,
+//           multiples of these next assignments may be serviced by
+//           the server threads that were previously assigned lighter
+//           workloads.
+//        4. Make subsequent scanning assignments as follows:
+//             a) 8 assignments of size 16 clusters
+//             b) 8 assignments of size 8 clusters
+//             c) 16 assignments of size 4 clusters
+//
+//    When there is no more remembered set processing work to be
+//    assigned to a newly idled worker threads, that thread can move
+//    on to work on other tasks associated with root scanning until such
+//    time as all clusters have been examined.
+//
+//  Once all clusters have been processed, the gang of GC worker
+//  threads collaborate to merge the overreach data.
+//
+//  for each ShenandoahHeapRegion old_region in the whole heap
+//    determine the cluster number of the first cluster belonging
+//      to that region
+//    for each cluster contained within that region
+//      Assure that exactly one worker thread initializes each
+//      cluster of overreach memory by invoking:
+//
+//        rs->merge_overreach(cluster_no, cluster_count)
+//
+//      in separate threads.  (Divide up the clusters so that
+//      different threads are responsible for merging different
+//      clusters.  Merging cost is essentially identical for
+//      each cluster.)
+//
+//
+// To initiate concurrent scanning, insert the above code sequences
+// into strong_roots_do and roots_do methods of ShenandoahRootScanner
+// (source file shenandoahRootProcessor.cpp) following the
+// construction of tc_cl and rm and before the call to
+// _serial_roots.oops_do().
+//
+// To initiate concurrent evacuation, insert the above code sequences
+// into ShenandoahRootEvacuator::roots_do() before the invocation of
+// _serial_roots.oops_do(oops, worker_id).
+
```

Comment:
This is not quite clear to me. Above you said that all of the work is done during safepoints. Here you mention 'concurrent scanning' ... what is true?